Add collision weight to `PhysicsBody2D/3D` for penetrations which must be avoided

TokageItLab opened the pull request:
Co-authored-by: Juan Linietsky <reduzio@gmail.com>

Fixes https://github.com/godotengine/godot-proposals/issues/3065

To handle penetrations that absolutely must be avoided, this PR implements `collision_weight` in the `PhysicsBody`.

For example, if a player is sandwiched between two collisions, the one with the smaller collision weight will be penetrated with priority. This prevents the player from being pushed by a smaller enemy and getting stuck by penetrating terrain.

https://user-images.githubusercontent.com/61938263/184455306-4997bbfb-9fb1-4e1c-aef9-0202bd1839c2.mp4

https://user-images.githubusercontent.com/61938263/184455322-ac6871e2-1889-4f1e-a43b-ec10be88b81b.mp4

[physics_test_2.zip](https://github.com/godotengine/godot/files/9330214/physics_test_2.zip)

TokageItLab committed:
Add collision weight to PhysicsBody for penetrations must be avoided

Co-authored-by: Juan Linietsky <reduzio@gmail.com>

fabriceci commented on doc/classes/CollisionObject2D.xml:
```diff
@@ -206,6 +206,9 @@
 			The physics layers this CollisionObject2D scans. Collision objects can scan one or more of 32 different layers. See also [member collision_layer].
 			[b]Note:[/b] Object A can detect a contact with object B only if object B is in any of the layers that object A scans. See [url=$DOCS_URL/tutorials/physics/physics_introduction.html#collision-layers-and-masks]Collision layers and masks[/url] in the documentation for more information.
 		</member>
+		<member name="collision_weight" type="float" setter="set_collision_weight" getter="get_collision_weight" default="1.0">
```
How about calling it "collision_priority" (or similar), the word weight can be confusing as a static body has infinite mass.

fabriceci commented on servers/physics_2d/godot_space_2d.cpp:
```diff
@@ -686,6 +692,12 @@ bool GodotSpace2D::test_body_motion(GodotBody2D *p_body, const PhysicsServer2D::
+			for (int i = 0; i < cbk.amount; i++) {
+				inv_total_weight += wt[i];
+			}
+			inv_total_weight = Math::is_zero_approx(inv_total_weight) ? 1.0 : 1.0 / inv_total_weight;
```
Rather than applying a recovery proportional to the weight, couldn't we exclude, for example at the second iteration, the collision shape with the lowest priority (to avoid the body, as seen in the first video, sinking a bit into the ground)?

fabriceci replied:
@reduz Do you have an opinion on this? It's more elegant the current solution, but if we exclude the recover for lower priority shape, it would avoid in the first video to entering the ground for a few frames and therefore, potentially recovering the information that the object is stuck in the ground (and also simply the behaviour -> recovers or not), I'm not sure.

TokageItLab commented:
@fabriceci Indeed, in the realm of physics, the name weight may conflict with mass, so I renamed it from weight to priority.

I also changed the motion on return to use bool instead of having weights, but I am not sure what side effects this may cause, so please test. Especially, I think we need to test it to see if there are any bugs, especially if there are more than 3 collisions, or if there is any penetration after jumping between 2 cylinders of different weights, etc.

@reduz Do you have any concerns about this `&& priorities[i] >= priority_max`?
```cpp
if (depth > min_contact_depth + CMP_EPSILON && priorities[i] >= priority_max) {
	// Only recover if there is penetration.
	recover_motion -= n * (depth - min_contact_depth) * 0.4;
}
```

TokageItLab commented:
@fabriceci @reduz I tested some cases. In conclusion, it is definitely better to use weight.

Below is a video of the test when sandwiched between cylinders with different weights:

Using bool:

https://user-images.githubusercontent.com/61938263/184559598-42b8f135-bbbc-4bf3-a69f-ad03761e5978.mp4

When using bool, it means that one side's penetration is not taken care of at all. As a result, penetration is more likely to occur and can easily get stuck.

Using weight:

https://user-images.githubusercontent.com/61938263/184559588-92bbd791-ce36-4eca-a874-9376376a5c9f.mp4

When using weight, all penetrations are taken care of with some bias, so it is possible to get out of the stack.

---

In any case, some special handling is often necessary when the object is caught in a collision.

If you just want to prevent bounce, you can set a large difference in weights. Also, if you need to completely ignore one of the collisions, I think the right way is to use `get_depth()`, etc. and implement unstack handling in user-side scripts.

TokageItLab commented:
As a result:
- Implement a property named `float collision_priority`
- Consider a weight dependent on `collision_priority` in the calculation of the recovered motion

fabriceci reviewed (dismissed):
Thanks for trying the suggestion! It seems perfect to me.

rburing commented:
My concern is that with the default `collision_priority = 1.0`, depenetration will be weakened if there is more than one collision: the whole `recover_motion` is multiplied by the inverse of the number of collisions. So the code changes the default physics behavior, and as such I think it needs more testing.

By the way, in the code the multiplication by `inv_total_weight` is done in each summand; to save some multiplications it could be done once at the end (unless this affects precision negatively; I don't know if it does).

TokageItLab commented:
@rburing 

> My concern is that with the default collision_priority = 1.0, depenetration will be weakened if there is more than one collision: the whole recover_motion is multiplied by the inverse of the number of collisions. So the code changes the default physics behavior, and as such I think it needs more testing.

I modified it to multiply by the amount.

```cpp
inv_total_weight = Math::is_zero_approx(inv_total_weight) ? 1.0 : 1.0 / inv_total_weight;
```
↓
```cpp
inv_total_weight = Math::is_zero_approx(inv_total_weight) ? 1.0 : 1.0 / inv_total_weight * (real_t)cbk.amount;
```

This makes bounce more likely to occur and increases the ratio to prevent bounce like `1:1000`. But, this should theoretically make the default recover motion value match the original calculation.

---

> By the way, in the code the multiplication by inv_total_weight is done in each summand; to save some multiplications it could be done once at the end (unless this affects precision negatively; I don't know if it does).

The same constant of 0.4 can also be moved out, but after trying it, the behavior is clearly bad. I don't know if it is a precision issue or not, but I don't touch it.

TokageItLab commented:
Normalized weights. Now, for example, the behavior should be the same when priority is 1:1 as when priority is 100:100.

fabriceci reviewed (commented):
I had redone a small review, I forgot to post it, I look at the last changes later.

TokageItLab commented on servers/physics_2d/godot_space_2d.cpp:
```diff
@@ -686,6 +692,12 @@ bool GodotSpace2D::test_body_motion(GodotBody2D *p_body, const PhysicsServer2D::
+			for (int i = 0; i < cbk.amount; i++) {
+				inv_total_weight += priorities[i];
+			}
+			inv_total_weight = Math::is_zero_approx(inv_total_weight) ? 1.0 : 1.0 / inv_total_weight;
```
Outdated, now it is multiplied by amount and normalized weight.
```
inv_max_weight = Math::is_zero_approx(inv_max_weight) ? 1.0 : 1.0 / inv_max_weight;
inv_total_weight = Math::is_zero_approx(inv_total_weight) ? 1.0 : (real_t)cbk.amount / (inv_total_weight * inv_max_weight) * inv_max_weight;
```

TokageItLab commented on scene/3d/collision_object_3d.cpp:
```diff
@@ -466,6 +479,7 @@ void CollisionObject3D::_bind_methods() {
 	ADD_GROUP("Collision", "collision_");
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "collision_layer", PROPERTY_HINT_LAYERS_3D_PHYSICS), "set_collision_layer", "get_collision_layer");
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "collision_mask", PROPERTY_HINT_LAYERS_3D_PHYSICS), "set_collision_mask", "get_collision_mask");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "collision_priority"), "set_collision_priority", "get_collision_priority");
```
At least zero or negative values are undesirable, but I don't think it needs to be hint_range, since any number of other values could be valid. Even a large ratio is needed to reduce bounce.

fabriceci replied:
What do you think about checking in the setter if the value is < 1 (and force it to 1)?  I am afraid that some people try to put a value of 0 (maybe we can also add the information in the description)

TokageItLab replied:
When there are many priority1 objects, there is a possibility to set a value less than 1 when placing a weaker object. So at least, the minimum value should be 0.001.

TokageItLab replied:
For now, only simple errors have been implemented. If the value is forced to be modified, the implementation will be complicated because the priority value exists in two separate places: the interface object and the actual object.

TokageItLab commented:
Fixed the parts pointed out.

rburing commented:
Multiplying by `cbk.amount` addresses my concern, keeping the defaults as they are.

What is the point of using the maximum? I don't get it. The `100 : 100` case should already be the same as `1 : 1` without it.

TokageItLab commented:
Hmm... indeed, maybe I did something wrong during the test. Removed `max_weight`.

TokageItLab commented:
@reduz Okay, ready for review.

TokageItLab commented:
Added error for a priority of 0 or less.

reduz commented:
Would be interesting to figure out how to do this with the regular physics solver, which should be totally possible.

akien-mga commented:
Thanks!

vermeeren commented:
@TokageItLab @akien-mga Could it be the setting for collision priority was forgotten for the `GridMap` node? I don't think setting the priority in the underlying mesh library meshes' `StaticBody3D` has effect once you export it into a mesh library.

![Screenshot_20221209_010649](https://user-images.githubusercontent.com/5957454/206592500-aa734abf-8aa7-4414-9430-e8b3aac597f9.png)

The world (walls, floor) is made by `GridMap` cubes. The player is the bean and there are 3 balls to push around. In this case, the balls collide with the player but the player does **not** collide with the balls. (i.e. player movement is never blocked).

My goal is to set the walls (`GridMap`) to a high priority like `10` to avoid the player with priority `1` pushing the balls through the wall, which is currently pretty easy to do in the corners even with such thick walls. Using f692b476869c14023b342c12efc21cb5cd6002b7 at the moment.

Perhaps this should be a new issue, but I'm not sure which is why I comment here first.
